src/server/pfs/fuse/server.go

```diff
@@ -0,0 +1,850 @@
+	// TCP port (just for convenient manual testing with curl for now...)
+	// TODO: make port and bind ip parameterizable
+	srv := &http.Server{Addr: ":9002", Handler: router}
+
+	go func() {
+		err := mm.Start()
+		if err != nil {
+			fmt.Printf("Error running mount manager: %s\n", err)
+			os.Exit(1)
+		}
+		err = mm.Cleanup()
+		if err != nil {
+			fmt.Printf("Error cleaning up mount manager: %s\n", err)
+			os.Exit(1)
+		}
+		srv.Shutdown(context.Background())
+	}()
+
+	return srv.ListenAndServe()
+}
+
+type MountState struct {
+	Name       string   `json:"name"`       // where to mount it. written by client
+	MountKey   MountKey `json:"mount_key"`  // what to mount. written by client
+	Mode       string   `json:"mode"`       // "ro", "rw", or "" if unknown/unspecified. written by client
+	State      string   `json:"state"`      // "unmounted", "mounting", "mounted", "pushing", "unmounted", "error". written by fsm
+	Status     string   `json:"status"`     // human readable string with additional info wrt State, e.g. an error message for the error state. written by fsm
+	Mountpoint string   `json:"mountpoint"` // where on the filesystem it's mounted. written by fsm. can also be derived from {MountDir}/{Name}
+}
+
+type MountStateMachine struct {
+	MountState
+	manager   *MountManager
+	mu        sync.Mutex
+	requests  chan Request
+	responses chan Response
+}
+
+// TODO: switch to pach internal types if appropriate?
+type ListResponse map[string]RepoResponse
+
+type BranchResponse struct {
+	Name  string     `json:"name"`
+	Mount MountState `json:"mount"`
+}
+
+type RepoResponse struct {
+	Name     string                    `json:"name"`
+	Branches map[string]BranchResponse `json:"branches"`
+	// TODO: Commits map[string]CommitResponse
+}
+
+type GetResponse RepoResponse
+
+type MountKey struct {
+	Repo   string
+	Branch string
+	Commit string
+}
+
+func (m *MountKey) String() string {
+	// m.Commit is optional
+	if m.Commit == "" {
+		return fmt.Sprintf("%s/%s", m.Repo, m.Branch)
+	} else {
+		return fmt.Sprintf("%s/%s/%s", m.Repo, m.Branch, m.Commit)
+	}
+}
+
+func mountKeyFromString(key string) (MountKey, error) {
+	shrapnel := strings.Split(key, "/")
+	if len(shrapnel) == 3 {
+		return MountKey{
+			Repo:   shrapnel[0],
+			Branch: shrapnel[1],
+			Commit: shrapnel[2],
+		}, nil
+	} else if len(shrapnel) == 2 {
+		return MountKey{
+			Repo:   shrapnel[0],
+			Branch: shrapnel[1],
+		}, nil
+	} else {
+		return MountKey{}, fmt.Errorf(
+			"wrong number of '/'s in key %+v, got %d expected 2 or 3", key, len(shrapnel),
+		)
+	}
+}
+
+// state machines in the style of Rob Pike's Lexical Scanning
+// http://www.timschmelmer.com/2013/08/state-machines-with-state-functions-in.html
+
+// setup
+
+type StateFn func(*MountStateMachine) StateFn
+
+func (m *MountStateMachine) transitionedTo(state, status string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	fmt.Printf("[%s] %s -> %s\n", m.MountKey, m.State, state)
```

Comment:
stray?

Reply:
This one's intentional, it's useful to be able to see the behavior of the system in terms of which state transitions it's performing.

Reply:
How hard would it be to switch to logrus for these?

Reply:
done! the output looks pretty now 😁 

![image](https://user-images.githubusercontent.com/264658/151032740-758c9dc4-6c77-4ff4-9f0c-4a18199fc4ce.png)